### PR TITLE
Fix issue where benchmark execution duration is incorrect in check. Closes #1249

### LINKED ISSUE #1249: `benchmark` duration in `check` is incorrect

### PR DESCRIPTION
**Pull request by binaek**

### CODE REVIEW

File: cmd/check.go [hunk 1]
```diff
@@ -174,7 +174,7 @@ func runCheckCmd(cmd *cobra.Command, args []string) {
 		err = displayControlResults(ctx, executionTree)
 		utils.FailOnError(err)
 
-		if len(exportFormats) > 0 {
+		if len(exportFormats) > 0 && !utils.IsContextCancelled(ctx) {
```
Comment (kaidaguerre) [inline cmd/check.go:177]: can you move `!utils.IsContextCancelled(ctx)` into the start of `exportCheckResult`

File: cmd/check.go [hunk 2]
```diff
@@ -288,8 +289,8 @@ func initialiseCheck(ctx context.Context, spinner *spinner.Spinner) *checkInitDa
 	// register EnsureSessionData as a callback on the client.
 	// if the underlying SQL client has certain errors (for example context expiry) it will reset the session
 	// so our client object calls this callback to restore the session data
-	initData.client.SetEnsureSessionDataFunc(func(ctx context.Context, conn *db_common.DatabaseSession) (error, []string) {
-		return workspace.EnsureSessionData(ctx, sessionDataSource, conn)
+	initData.client.SetEnsureSessionDataFunc(func(lCtx context.Context, conn *db_common.DatabaseSession) (error, []string) {
```
Comment (kaidaguerre) [inline cmd/check.go:292]: lCtx??

File: control/controlexecute/control_run.go [hunk 3]
```diff
@@ -147,24 +169,31 @@ func (r *ControlRun) Execute(ctx context.Context, client db_common.Client) {
...
 		r.Lifecycle.Add("execute_end")
-		r.Duration = r.Lifecycle.GetDuration()
+		r.Duration = time.Since(startTime)
+		if r.group != nil {
+			r.group.addDuration(r.Duration)
+		}
 		log.Printf("[TRACE] finishing with concurrency, %s, , %d\n", r.Control.Name(), r.executionTree.progress.executing)
 	}()
 
+	if utils.IsContextCancelled(ctx) {
```
Comment (kaidaguerre) [inline control/controlexecute/control_run.go:179]: why do we need this here? how did it work before?

File: control/controlexecute/result_group.go [hunk 4]
```diff
@@ -49,12 +50,13 @@ func NewGroupSummary() *GroupSummary {
...
-		Tags:              make(map[string]string),
-		Summary:           NewGroupSummary(),
-		Severity:          make(map[string]StatusSummary),
-		summaryUpdateLock: new(sync.Mutex),
+		GroupId:            RootResultGroupName,
+		Groups:             []*ResultGroup{},
+		Tags:               make(map[string]string),
+		Summary:            NewGroupSummary(),
+		Severity:           make(map[string]StatusSummary),
+		durationUpdateLock: new(sync.Mutex),
```
Comment (kaidaguerre) [inline control/controlexecute/result_group.go:58]: do we really need 2 different locks?

File: db/db_client/db_client_session.go [hunk 5]
```diff
@@ -14,11 +14,17 @@ import (
 	"github.com/turbot/steampipe/utils"
 )
 
-func (c *DbClient) AcquireSession(ctx context.Context) *db_common.AcquireSessionResult {
+func (c *DbClient) AcquireSession(ctx context.Context) (asr *db_common.AcquireSessionResult) {
```
Comment (kaidaguerre) [inline db/db_client/db_client_session.go:17]: please do not use acronyms for variable names - `res` will do

Comment (kaidaguerre) [reply db/db_client/db_client_session.go:17]: I am often having to mention this - please remember for the future - no abbreviations/acronyms

File: db/db_common/db_session.go [hunk 6]
```diff
@@ -17,6 +18,7 @@ type DatabaseSession struct {
 	LastUsed    time.Time             `json:"last_used"`
 	SearchPath  []string              `json:"-"`
 	Initialized bool                  `json:"-"`
+	inUse       int32                 `json:"-"`
```
Comment (kaidaguerre) [inline db/db_common/db_session.go:21]: what is this for?

**Review by kaidaguerre — CHANGES_REQUESTED** [review]
makes mer nervous to introduce multiple new sync flags - do we really need them all? can we refactor to reduce this?

File: control/controlexecute/control_run.go (hunk 3 above)
Comment (binaek) [reply control/controlexecute/control_run.go:179]: earlier, the ones which were waiting for connection  would report a `error acquiring database connection` error.

with this change, it reports the proper `execution cancelled` error

Comment (binaek) [reply control/controlexecute/control_run.go:179]: my mistake - there's no point in putting it there

**Pull request merged**